torch/csrc/jit/tensorexpr/kernel.cpp

```diff
@@ -670,29 +670,40 @@ bool loopBoundsAllEqual(const std::vector<ForPtr>& loops) {
 // indices where none would be needed, which would significantly complicate
 // vectorization.
 void fuseAllLoops(StmtPtr st) {
-  if (auto block = to<tensorexpr::Block>(st)) {
-    std::vector<ForPtr> loopsToFuse;
-    for (auto stmt : *block) {
-      auto loop = to<For>(stmt);
-      if (!loop) {
```

Comment:
It's concerning to me that this check, which aborts loop fusion if a non-loop statement is encountered, has been removed in this diff.  Specifically I'm concerned about a situation where there is a memory dependence, e.g.:
```
for i : (0, M)
  b[i] = a[i]
b[0] = 12
for i : (0, M)
  c[i] = b[i]
```
If we naively fuse those two loops, we'll miss the non-loop update of `b[0]` in line 3.  Now, it's possible that loop fusion itself will prevent this case (does it reject fusion of non-adjacent loops?  I can't remember).

In any case, please either point me to some existing tests that demonstrate correctness in such a case, write new tests exercising that behavior, or restore a check refusing to fuse things with non-loops.

Reply:
This loop enumerates the blocks within a given block. If the current block is not for-statement, the fusion would be broken.  Take the code snippet as an example, `b[0] = 12` is not a for-statement, hence the two loops will not be fused.

Reply:
This PR targets optimizing the loop parallelism. The original logic would not parallelize the first and second loop if the buffer is not the output of a fusion group.

Reply:
@bertmaher , this [UT](https://github.com/pytorch/pytorch/blob/master/test/cpp/tensorexpr/test_loopnest.cpp#L5613) is to test the case. But I polished it a little bit by changing the loop interval to make sure the two loop intervals are the same but only a store is inserted between the two loops.

https://github.com/pytorch/pytorch/pull/85056/files#diff-27612a2830a19cb267524fa9e43a085d1c2de71a987754034da953c1c16b99e6R5628